### PR TITLE
AAF-95 Write Routes (#95)

### PR DESCRIPTION
**Pull request by jwu910**
Added bulk of the write routes.

TODO:
Write tests
Error Handling

### CODE REVIEW

File: server/routes/api.js [hunk 1]
```diff
@@ -115,31 +119,40 @@ routes.get('/pairing/paired', (req, res) => {
   });
 });
 
-routes.get('/wishlist/:familyId', (req, res) => {
+routes.patch('/wishlist/:familyId/addItem', wishlistController.addItem);
+
+routes.patch('/wishlist/:familyId/create', wishlistController.create);
```
Comment (rainrivas) [inline server/routes/api.js:124]: consider making this a `PUT` to imply creation of a new object.
> The HTTP PUT method only allows complete replacement of a document. Unlike PUT, PATCH is not idempotent, meaning successive identical patch requests may have different effects. However, it is possible to issue PATCH requests in such a way as to be idempotent.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH

File: server/routes/api.js [hunk 2]
```diff
@@ -115,31 +119,40 @@ routes.get('/pairing/paired', (req, res) => {
...
+
+routes.patch('/wishlist/:familyId/removeItem', wishlistController.removeItem);
+
+routes.patch('/wishlist/:familyId/updateItem', wishlistController.updateItem);
 
+routes.get('/wishlist/:familyId', (req, res) => {
```
Comment (rainrivas) [inline server/routes/api.js:130]: from an optimization standpoint, I *think* this route should go before the routes built off this, e.g. `/wishlist/:familyId/updateItem`

File: server/controllers/wishlistController.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  // Single items should still reside in an Array
+
+  try {
+    Wishlist.where({ family: familyId }).exec((err, wishlist) => {
+      if (err) {
+        res.status(500).json({ Error: err });
+      }
+
+      if (wishlist[0].list !== []) {
+        return res.status(200).json({
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:76]: Doesn't make sense for this to be a 200 response, possibly use a 409 or 412 response

File: server/controllers/wishlistController.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        });
+      });
+    });
+  } catch (err) {
+    res.status(500).json({ Message: 'TEMP ERROR TEST' });
+  }
+};
+
+exports.delete = (req, res) => {
+  res.send('NOT YET IMPLEMENTED');
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:102]: send a 501 status code

File: server/controllers/wishlistController.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        'list._id': itemId,
+      })
+      .exec((err, wishlist) => {
+        if (err) {
+          res.status(500).json({ Error: err });
+        }
+
+        if (!wishlist) {
+          return res
+            .status(400)
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:130]: Your error states 'not found', should we set this to 404?

File: server/controllers/wishlistController.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        );
+
+        wishlist.totalListCost = _getWishlistCost(wishlist.list);
+
+        wishlist.save((err, wishlist) => {
+          if (err) {
+            res.status(500).json({ Error: err });
+          }
+
+          res.status(200).json({
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:145]: from MDN: 
> The successful result of a PUT or a DELETE is often not a 200 OK but a 204 No Content (or a 201 Created when the resource is uploaded for the first time).

Also, don't forget to put he body of newly created/modified list in the response

File: server/controllers/wishlistController.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+      wishlist[0].list.push(item);
+
+      wishlist[0].totalListCost = _getWishlistCost(wishlist[0].list);
+
+      wishlist[0].save((err, wishlist) => {
+        if (err) {
+          res.status(500).json({ Error: err });
+        }
+
+        res.status(201).json({
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:44]: You should be returning the newly created object in the response.
From MDN: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201
The HTTP 201 Created success status response code indicates that the request has succeeded and has led to the creation of a resource. **The new resource is effectively created before this response is sent back. and the new resource is returned in the body of the message**, its location being either the URL of the request, or the content of the Location header

File: server/controllers/wishlistController.js [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+      wishlist[0].list = newWishlist;
+
+      wishlist[0].totalListCost = _getWishlistCost(wishlist[0].list);
+
+      wishlist[0].save((err, wishlist) => {
+        if (err) {
+          res.status(500).json({ Error: err });
+        }
+
+        res.status(200).json({
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:90]: Add the wishlist to response

File: server/controllers/wishlistController.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        });
+
+        wishlist.totalListCost = _getWishlistCost(wishlist.list);
+
+        wishlist.save((err, wishlist) => {
+          if (err) {
+            res.status(500).json({ Error: err });
+          }
+
+          res.status(200).json({
```
Comment (rainrivas) [inline server/controllers/wishlistController.js:193]: set to 201 and return the new list in the response

**Review by rainrivas — CHANGES_REQUESTED** [review]
Fixup status codes and add wishlist to responses

File: server/controllers/wishlistController.js (hunk 4 above)
Comment (jwu910) [reply server/controllers/wishlistController.js:102]: Oh yeah these were place holders for the other route logic. I can udpate, and Duc can add his logic in when he refactors.

File: server/controllers/wishlistController.js (hunk 6 above)
Comment (jwu910) [reply server/controllers/wishlistController.js:145]: Oh okay, return the body of what was added? 

Currently, these routes are just returning the entire new family wishlist. I can change it.

Comment (rainrivas) [reply server/controllers/wishlistController.js:145]: @jwu910 I believe returning the entire new thing is the norm. so even if you had a wishlist with 10 items and you added item 11, you would return all 11 items and not just the 1 change

File: server/controllers/wishlistController.js (hunk 4 above)
Comment (rainrivas) [reply server/controllers/wishlistController.js:102]: I dont think anyone is working on delete at the moment, so just set it to 501 until someone else implements stuff

Comment (jwu910) [reply server/controllers/wishlistController.js:102]: I think @psyduc finished it already...

File: server/controllers/wishlistController.js (hunk 6 above)
Comment (jwu910) [reply server/controllers/wishlistController.js:145]: @rainrivas Oh we're good then. I'll just make sure the status codes are updated.

File: server/controllers/wishlistController.js (hunk 8 above)
Comment (jwu910) [reply server/controllers/wishlistController.js:90]: Wishlist is already included in this response. Was this just overlooked, or did you want it formatted differently?

Comment (rainrivas) [reply server/controllers/wishlistController.js:90]: @jwu910 I didn't see wishlist here the first time I read through this. you can disregard this one

File: server/controllers/wishlistController.js (hunk 5 above)
Comment (rainrivas) [reply server/controllers/wishlistController.js:130]: Also lets set this to one line to keep it consistent with all the other responses, like `res.status(500).json({ Error: err });` (ln 124)

**Comment by rainrivas** [comment]
@jwu910 any updates on this ?

**Comment by jwu910** [comment]
Not yet. Need to update tests still. Routes are working, but I don't think they catch all fail errors and I might need some help again. But I need to reassess where I left off.

**Comment by jwu910** [comment]
Gotta have to catch up with the latest code and get tests up.

Looks like we've got a conflict in routes/api I need to resolve too...

**Pull request closed without merging**